// components/Schedule/Days.js
import React, { Component } from 'react'
import moment from 'moment'
import { keys } from 'ramda'
import { Collection } from 'react-virtualized'

const WIDTH = 1512

class Days extends Component {
  weekDayCellRenderer = ({ key, style, index }) => {
    const day = moment().startOf('day').add(index, 'days').format('ddd M/DD')

    return (
      <div
        style={style}
        key={`${day}${index}`}
        className="br b--light-gray top-0 h-100"
      >
        <div className="bg-white w-100 tc pa2 bb b--light-gray">
          {day}
        </div>
      </div>
    )
  }

  weekDayCellSizeAndPositionGetter = ({ index }) => ({
    height: this.props.height,
    width: 216,
    x: 216 * index,
    y: 0
  })

  weekCellCount = () => Math.ceil(this.props.scrollWidth / 216)

  render() {
    return (
      <Collection
        scrollLeft={this.props.scrollLeft}
        className="br b--light-gray"
        cellRenderer={this.weekDayCellRenderer}
        cellSizeAndPositionGetter={this.weekDayCellSizeAndPositionGetter}
        height={this.props.height}
        width={WIDTH}
        cellCount={}
      />
    )
  }
}

export default Days
