// components/Schedule/Event.js
import React, { Component } from 'react'
import Draggable from 'react-draggable'
import moment from 'moment'
import TimeBox from './TimeBox'
import { getHourCount, getPixelCount } from './util'

class Event extends Component {
  constructor(props) {
    super(props)
    this.isDragging = false
    this.wasMouseDown = false
    this.state = {
      extendDir: null,
      extendAmt: 0,
      isSelfDragging: false
    }
  }

  isNearTop = () => this.props.translateY - 45 <= this.props.scrollTop

  handleMouseDown = () => {
    this.wasMouseDown = true
    setTimeout(() => {
      this.wasMouseDown = false
    }, 500)
  }

  handleMouseUp = () => {
    if (!this.isDragging && this.wasMouseDown) {
      this.wasMouseDown = false
      this.props.onClick(this.props.event)
    }
  }

  handleStart = extendDir => () => {
    this.props.handleStart(this.props.event._id)
    this.setState({ extendDir, isSelfDragging: true })
  }

  handleDrag = extendDir => (_, drag) => {
    if (!this.isDragging) {
      this.props.handleStart(this.props.event._id)
      this.setState({ extendDir, isSelfDragging: true })
    }
    this.isDragging = true
    this.props.handleDrag(drag.x)
    this.setState({ extendDir, extendAmt: drag.x, isSelfDragging: true })
  }

  handleStop = extendDir => (event, drag) => {
    setTimeout(() => (this.isDragging = false), 0)
    this.props.handleStop(extendDir)(event, drag, this.props.event)
    this.setState({ extendDir: null, extendAmt: 0, isSelfDragging: false })
  }

  getWidth = styleWidth => {
    if (!this.state.extendDir) return styleWidth
    if (this.state.extendDir === 'END') return styleWidth + this.state.extendAmt
    if (this.state.extendDir === 'START') {
      return styleWidth - this.state.extendAmt
    }
  }

  getLeft = styleLeft => {
    if (!this.state.extendDir) return styleLeft
    if (this.state.extendDir === 'END') return styleLeft
    if (this.state.extendDir === 'START') {
      return styleLeft + this.state.extendAmt
    }
  }

  render() {
    const { event, style, handleStop, translateY } = this.props
    const left = this.getLeft(style.left)
    const width = this.getWidth(style.width)
    const isUnfocused = this.props.isDragging && !this.state.isSelfDragging
    const opacity = 1
    const hourOffset = getHourCount(this.state.extendAmt)
    const eventPixelCount =
      getPixelCount(
        event.rangeStart,
        moment(event.rangeEnd).subtract(1, 'hour')
      ) - 10

    const fromTodayPixelCount =
      -1 *
      getPixelCount(
        moment().startOf('day').isSameOrBefore(event.rangeStart)
          ? moment().startOf('day')
          : moment(event.rangeStart),
        moment(event.rangeStart)
      )

    const maxHourOffset = getHourCount(fromTodayPixelCount)
    const moveToTop = () => this.props.moveToTop(event._id)

    return (
      <div
        className="relative"
        style={{
          transform: `translate3d(0, ${translateY}px, 0)`,
          transition: 'transform 0.3s',
          willChange: 'transform',
          zIndex: this.state.isSelfDragging ? 1 : 0
        }}
        onMouseDown={this.handleMouseDown}
        onMouseUp={this.handleMouseUp}
      >
        <Draggable
          axis="x"
          cancel="#extend"
          onStop={this.handleStop()}
          onDrag={this.handleDrag()}
          position={{ x: 0, y: 0 }}
          bounds={{ left: fromTodayPixelCount }}
        >
          <div>
            <div
              key={event._id}
              id={event._id}
              className="relative bg-light-gray br3 pt3 pl3"
              style={{
                ...style,
                width,
                left
              }}
            >
              <div className="ml1 relative">
                <div className="f5 truncate">
                  {moment(event.rangeStart)
                    .add(
                      this.state.extendDir === 'END'
                        ? 0
                        : Math.max(hourOffset, maxHourOffset),
                      'hours'
                    )
                    .format('MMMM D @ h A')}{' '}
                  to{' '}
                  {moment(event.rangeEnd)
                    .add(
                      this.state.extendDir === 'START'
                        ? 0
                        : Math.max(hourOffset, maxHourOffset),
                      'hours'
                    )
                    .format('MMMM D @ h A')}
                </div>
                <div className="f6 lh-copy truncate">
                  {this.props.describeEvent(event)}
                </div>
              </div>
            </div>
            <div className="" style={{ ...style, transition: 'top 0.5s' }}>
              <div id="extend">
                <Draggable
                  axis="x"
                  position={{ x: 0, y: 0 }}
                  onDrag={this.handleDrag('START')}
                  onStop={this.handleStop('START')}
                  bounds={{
                    right: eventPixelCount,
                    left: fromTodayPixelCount
                  }}
                >
                  <div
                    style={{ cursor: 'ew-resize', width: '3px' }}
                    className="absolute h-100 left-0 top-0 br3 br--left"
                  >
                    {this.state.extendDir === 'START' &&
                      <div>
                        <div
                          className="absolute"
                          style={{
                            height: '120%',
                            width: '2px',
                            top: '-11%',
                            left: '-1px',
                            backgroundColor: '#979797'
                          }}
                        />
                        <div
                          className="absolute"
                          style={{
                            right: '-40px',
                            top: this.isNearTop() ? '85px' : '-45px'
                          }}
                        >
                          <TimeBox
                            pointUp={this.isNearTop()}
                            time={moment(event.rangeStart)
                              .add(
                                this.state.extendDir === 'START'
                                  ? hourOffset
                                  : 0,
                                'hours'
                              )
                              .format('h A')}
                          />
                        </div>
                      </div>}
                  </div>
                </Draggable>
              </div>
              <div id="extend">
                <Draggable
                  axis="x"
                  position={{ x: 0, y: 0 }}
                  onDrag={this.handleDrag('END')}
                  onStop={this.handleStop('END')}
                  bounds={{ left: -1 * eventPixelCount }}
                >
                  <div
                    style={{ cursor: 'ew-resize', width: '3px' }}
                    className="absolute h-100 right-0 top-0 br3 br--right"
                  >
                    {this.state.extendDir === 'END' &&
                      <div>
                        <div
                          className="absolute"
                          style={{
                            height: '120%',
                            width: '2px',
                            top: '-11%',
                            right: '-1px',
                            backgroundColor: '#979797'
                          }}
                        />
                        <div
                          className="absolute"
                          style={{
                            left: '-39px',
                            top: this.isNearTop() ? '85px' : '-45px'
                          }}
                        >
                          <TimeBox
                            pointUp={this.isNearTop()}
                            time={moment(event.rangeEnd)
                              .add(
                                this.state.extendDir === 'END' ? hourOffset : 0,
                                'hours'
                              )
                              .format('h A')}
                          />
                        </div>
                      </div>}
                  </div>
                </Draggable>
              </div>
            </div>
          </div>
        </Draggable>
      </div>
    )
  }
}

export default Event
