// components/Schedule/index.js
import React, { Component } from 'react'
import { invoker, identity, times } from 'ramda'
import { ScrollSync } from 'react-virtualized'
import moment from 'moment'
import { getHourCount, getPixelCount } from './util'
import DefaultDropdown from './DefaultDropdown'
import DaysEvents from './DaysEvents'

class Schedule extends Component {
  constructor(props) {
    super(props)
    this.state = {
      scrollLeft: undefined,
    }
  }

  update = () => this.daysEvents.updateCollection()

  goTo = pixelCount =>
    this.setState({ scrollLeft: pixelCount }, () =>
      this.setState({ scrollLeft: undefined }),
    )

  scrollToTime = unit => (num, currentDate) => {
    const a = moment().startOf('day')
    const b = invoker(1, unit)(num, currentDate.startOf('month'))
    const c = moment(a).add(1, 'year')
    const d = b.isBefore(c) ? b : c
    const pixelDiff = getPixelCount(a, d)
    this.goTo(pixelDiff < 0 ? 0 : pixelDiff)
  }

  render() {
    const { events, onUpdate, onEventClick, describeEvent } = this.props
    const Dropdown = this.props.Dropdown || DefaultDropdown
    const Button = this.props.Button || <button />

    return (
      <ScrollSync>
        {({ scrollTop, scrollLeft, onScroll, scrollWidth, scrollHeight }) => {
          const currentDate = moment()
            .startOf('day')
            .add(getHourCount(scrollLeft), 'hours')
          const monthOptions = moment.months().map((month, idx) => ({
            value: idx,
            key: idx,
            text: month,
            disabled:
              moment(currentDate)
                .month(idx)
                .isAfter(
                  moment()
                    .startOf('month')
                    .add(1, 'year'),
                ) ||
              moment(currentDate)
                .month(idx)
                .isBefore(moment().startOf('day')),
          }))
          const yearOptions = times(identity, 2).map(idx => {
            const year = moment().year() + idx
            return {
              value: year,
              key: year,
              text: year,
            }
          })

          const monthProps = {
            value: currentDate.month(),
            onChange: (_, { value }) =>
              this.scrollToTime('month')(value, currentDate),
            options: monthOptions,
          }
          const todayProps = {
            onClick: () => this.goTo(0),
          }
          const yearProps = {
            value: currentDate.year(),
            onChange: (_, { value }) =>
              this.scrollToTime('year')(value, currentDate),
            options: yearOptions,
          }
          return (
            <div>
              <div className="tc mb3 flex flex-row justify-center">
                <div className="ph2">
                  {typeof Button === 'function' ? (
                    <Button {...todayProps}>Today</Button>
                  ) : (
                    React.cloneElement(Button, todayProps, 'Today')
                  )}
                </div>
                <div className="ph2">
                  {typeof Dropdown === 'function' ? (
                    <Dropdown {...monthProps} />
                  ) : (
                    React.cloneElement(Dropdown, monthProps)
                  )}
                </div>
                <div className="ph2">
                  {typeof Dropdown === 'function' ? (
                    <Dropdown {...yearProps} />
                  ) : (
                    React.cloneElement(Dropdown, yearProps)
                  )}
                </div>
              </div>
              <div
                className="relative center ba b--light-gray br3 overflow-hidden"
                style={{ width: '100%', maxWidth: '1500px', height: '600px' }}
              >
                <div className="absolute left-0 top-0 h-100 w-100">
                  <DaysEvents
                    ref={daysEvents => (this.daysEvents = daysEvents)}
                    events={events}
                    onScroll={onScroll}
                    onUpdate={onUpdate}
                    describeEvent={describeEvent}
                    onEventClick={onEventClick}
                    actualScrollLeft={scrollLeft}
                    scrollLeft={this.state.scrollLeft}
                    scrollTop={scrollTop}
                    scrollHeight={scrollHeight}
                    scrollWidth={scrollWidth}
                    height={scrollHeight}
                  />
                </div>
              </div>
            </div>
          )
        }}
      </ScrollSync>
    )
  }
}

export default Schedule
